File: boj/14502.py
# 출처 : https://www.acmicpc.net/problem/14502

from copy import deepcopy
from itertools import combinations
from collections import deque

# 전염이 완료된 후, 0의 갯수를 구하는 함수


def countZero(copyGraph):
    count = 0
    for i in range(n):
        for j in range(m):
            if(copyGraph[i][j] == 0):
                count += 1
    return count


n, m = map(int, input().split(' '))
answer = []
graph = []
for i in range(n):
    graph.append(list(map(int, input().split(' '))))

# 0이 위치한 좌표를 구해서 벽을 세울 수 있는 모든 조합을 만든다.
zeroList = []
for i in range(n):
    for j in range(m):
        if(graph[i][j] == 0):
            zeroList.append([i, j])
# print(zeroList)
combinations = (list(combinations(zeroList, 3)))

# 상하좌우를 위한
deltaList = [[-1, 0], [1, 0], [0, 1], [0, -1]]

# 1. 조합중에 하나 선택해서 벽을 세우고, 2. 전염병 돌리고, 3. 0의 갯수 체크
for combi in combinations:
    visit = [[0 for i in range(m)] for j in range(n)]
    copyGraph = deepcopy(graph)

    # 조합에서 하나 골라서 벽세우는 과정
    for point in combi:
        copyGraph[point[0]][point[1]] = 1

    # bfs로 전염병 돌리기
    queue = deque()
    for x in range(n):
        for y in range(m):
            if(copyGraph[x][y] == 2):
                visit[x][y] = 1
                queue.append([x, y])
    while(queue):
        curPoint = queue.popleft()
        copyGraph[curPoint[0]][curPoint[1]] = 2
        visit[curPoint[0]][curPoint[1]] = 1
        x = curPoint[0]
        y = curPoint[1]
        for d in deltaList:
            if(x+d[0] >= 0 and x+d[0] < n and y+d[1] >= 0 and y+d[1] < m):
                if(copyGraph[x+d[0]][y+d[1]] == 0):
                    queue.append([x+d[0], y+d[1]])

    # 전염병이 돌고나서 0의 갯수 체크
    answer.append(countZero(copyGraph))
print(max(answer))

File: js/14502.js
// 출처 : https://www.acmicpc.net/problem/14502
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const input = [];
rl.on("line", function (line) {
    input.push(line)
}).on("close", function () {
    solution(input);
    process.exit();
});

const getCombinations = function (arr, selectNumber) {
    const results = [];

    if (selectNumber === 1) return arr.map((value) => [value]);

    arr.forEach((fixed, index, origin) => {
        const rest = origin.slice(index + 1);
        const combinations = getCombinations(rest, selectNumber - 1);
        const attached = combinations.map((combination) => [fixed, ...combination]);
        results.push(...attached);
    })

    return results;
}


function solution(input) {
    const n = input[0].split(' ')[0]
    const m = input[0].split(' ')[1]
    let answers = []
    let answer = 0;
    let graph = []
    for (let i = 1; i < input.length; i++) {
        let l = input[i].split(' ')
        for (let j = 0; j < l.length; j++) {
            l[j] = parseInt(l[j]);
        }
        graph.push(l)
    }
    const arr = []
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (graph[i][j] == 0) {
                arr.push([i, j])
            }
        }
    }
    const combArr = getCombinations(arr, 3)

    // 모든 조합에 대해 전염병 돌리기
    for (let i = 0; i < combArr.length; i++) {
        let combi = combArr[i];
        let copyGraph = JSON.parse(JSON.stringify(graph))

        // 벽세움
        for (let j = 0; j < combi.length; j++) {
            // console.log(combi[j])
            copyGraph[combi[j][0]][combi[j][1]] = 1
        }
        // 
        let queue = []
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < m; j++) {
                if (copyGraph[i][j] == 2) {
                    queue.push([i, j])
                }
            }
        }
        // console.log("!!");
        // console.log(copyGraph);
        // console.log(queue);
        let moves = [[-1, 0], [1, 0], [0, -1], [0, 1]];
        while (queue.length > 0) {
            position = queue.shift();
            // console.log(position)
            x = position[0];
            y = position[1];
            copyGraph[x][y] = 2
            moves.forEach(move => {
                if (x + move[0] >= 0 && x + move[0] < n && y + move[1] >= 0 && y + move[1] < m && copyGraph[x + move[0]][y + move[1]] == 0) {
                    copyGraph[x + move[0]][y + move[1]] = 2;
                    queue.push([x + move[0], y + move[1]]);
                }
            });
        }

        // console.log(graph)
        // console.log(copyGraph)
        // 0개수 세기
        let count = 0;
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < m; j++) {
                if (copyGraph[i][j] == 0) {
                    count++;
                }
            }
        }
        // answers.push(count)
        answer = Math.max(count, answer)
    }


    console.log(answer)
}